tree-walk.c

```diff
@@ -547,13 +562,25 @@ int get_tree_entry(const struct object_id *tree_oid, const char *name, struct ob
 		retval = -1;
 	} else {
 		struct tree_desc t;
-		init_tree_desc(&t, tree, size);
-		retval = find_tree_entry(&t, name, oid, mode);
+		if (gently) {
+			if (init_tree_desc_gently(&t, tree, size)) {
+				free(tree);
+				return -1;
```

Comment:
I think it would be better to set `retval` to -1 here, and then `goto done`.  If at some point this function grows, and we allocate some more memory before here, when we make the change we would probably forget to free it.  Although I see we do things the same way in line 557-558, so maybe that's not a valid concern.

Reply:
It makes sense. I am not sure what is git org opinion on using `goto` statements. In most cases, such a statement would make the code more unreadable; but I think in this case it would be good alternative and actually make the code future-proof.

Reply:
```
$ git grep goto | wc -l
1032
```
While there may be a few false positives in there, from a quick look it looks like mostly these are `goto` statements.  So I'd say it's fairly widespread, and I do feel like it would make sense to use it here.

Reply:
It *is* fairly common to use `goto` in Git's source code to simplify resource management (`free`, `close`) in error code paths.